runelite-client/src/main/java/net/runelite/client/plugins/hydra/HydraPlugin.java

```diff
@@ -0,0 +1,424 @@
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package net.runelite.client.plugins.hydra;
+
+import com.google.inject.Provides;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import lombok.AccessLevel;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+import net.runelite.api.Actor;
+import net.runelite.api.Client;
+import net.runelite.api.NPC;
+import net.runelite.api.Player;
+import net.runelite.api.Projectile;
+import net.runelite.api.ProjectileID;
+import net.runelite.api.coords.LocalPoint;
+import net.runelite.api.coords.WorldPoint;
+import net.runelite.api.events.AnimationChanged;
+import net.runelite.api.events.InteractingChanged;
+import net.runelite.api.events.NpcDespawned;
+import net.runelite.api.events.NpcSpawned;
+import net.runelite.api.events.ProjectileMoved;
+import net.runelite.client.config.ConfigManager;
+import net.runelite.client.eventbus.Subscribe;
+import net.runelite.client.events.ConfigChanged;
+import net.runelite.client.plugins.Plugin;
+import net.runelite.client.plugins.PluginDescriptor;
+import net.runelite.client.plugins.PluginType;
+import net.runelite.client.ui.overlay.OverlayManager;
+
+@PluginDescriptor(
+	name = "Hydra Helper",
+	description = "Overlays for normal Hydras.",
+	tags = {"hydra", "helper", "baby", "small", "normal", "regular"},
+	type = PluginType.PVM,
+	enabledByDefault = false
+)
+@Slf4j
+@Singleton
+public class HydraPlugin extends Plugin
+{
+	private static final Set<HydraAnimation> VALID_HYDRA_ANIMATIONS = Set.of(
+		HydraAnimation.RANGE,
+		HydraAnimation.MAGIC
+	);
+
+	@Getter(AccessLevel.PACKAGE)
+	private Map<Integer, Hydra> hydras = new HashMap<>();
+
+	@Getter(AccessLevel.PACKAGE)
+	private Map<LocalPoint, Projectile> poisonProjectiles = new HashMap<>();
+
+	@Getter(AccessLevel.PACKAGE)
+	private NPC interactingNpc = null;
+
+	private boolean attackCounterOverlay = false;
+
+	@Getter(AccessLevel.PACKAGE)
+	private boolean boldAttackCounterOverlay = false;
+
+	private boolean prayerOverlay = false;
+
+	private boolean poisonOverlay = false;
+
+	@Getter(AccessLevel.PACKAGE)
+	@Inject
+	private Client client;
```

Comment:
Inject a client reference into the other classes as well instead, this doesn't really matter but it looks better that way imo

Reply:
I've been meaning to look into how these annotations work (Inject, Singleton, Provides in particular). I'm guessing inject just adds an initialization into the constructor(s)? Is there any difference putting it on a class field vs the constructor+parameter? Does this create multiple client objects?

Clearly I need to look into it more xd

Reply:
`@Inject` creates/injects instances of objects. There's a bit of discussion (like in general) about ctor/field injection, but the way it's mostly done in the repo is use field injection on depending objects (plugins), and use constructor injection on dependant objects (overlays etc)

The singleton annotation is used a bit too much, cause it only means javax won't create multiple instances of the class. Kinda useless, seeing as how javax doesn't create instances of plugins.